### ISSUE TITLE
ModelCheckpoint monitor should be optional

### ISSUE THREAD

**Issue by carmocca** [PyTorchLightning/pytorch-lightning#3620]
## 🚀 Feature

We should be able to set `monitor = None`, meaning that it would always save

### Motivation

Current behaviour forces you to have a monitor, so to always save you would have to monitor something like the current epoch.

**Comment by williamFalcon** [comment]
how would you know when to save without monitoring something?

I personally wouldn't just want a checkpoint saved on every epoch

**Comment by carmocca** [comment]
Maybe just by itself is a bit useless, but combined with `period` one can have a "every_k_epochs" `ModelCheckpoint`

**Comment by williamFalcon** [comment]
well, generally checkpoints correspond to some sort of improvement in performance...

there is a (rare) case (image synthesis, speech synthesis), etc... where longer training does make more sense, in which case saving the last checkpoint makes sense.

in this case though, the answer is just that whenever save last is enabled, it always saves without having to monitor something?

There is something to be said about the default behavior of saving checkpoint on every epoch by default until the user monitors a specific metric. that might not be a bad option (thoughts @ananthsub ?)

**Comment by ananthsub** [comment]
Couldn't you set `save_top_k = -1`? Then `monitor` is ignored

**Comment by carmocca** [comment]
Not really, because you still would have to give a placeholder monitor.

Currently, if monitor is None, this line fails:

https://github.com/PyTorchLightning/pytorch-lightning/blob/c61e1e697d2e4ed5cc97c187e6704a44f2f77aff/pytorch_lightning/callbacks/model_checkpoint.py#L177

And if you give it a placeholder monitor `foo` (even with `save_top_k = -1`) it has to be present in the logged metrics. see:

https://github.com/PyTorchLightning/pytorch-lightning/blob/c61e1e697d2e4ed5cc97c187e6704a44f2f77aff/pytorch_lightning/callbacks/model_checkpoint.py#L340-L345

https://github.com/PyTorchLightning/pytorch-lightning/blob/c61e1e697d2e4ed5cc97c187e6704a44f2f77aff/pytorch_lightning/callbacks/model_checkpoint.py#L414-L415

**Comment by carmocca** [comment]
> well, generally checkpoints correspond to some sort of improvement in performance...

They do correspond to that right now, but we should see them just as snapshot takers.

When we support having multiple checkpoint callbacks, if you are training for a long time you might want to take snapshots of your model not only on performance improvements but also every N epochs (regardless of performance). For example, to have an history of how the model improved.

This issue is just one step forward to support that.

**Comment by awaelchli** [comment]
@carmocca with your None approach, what will the value of best_model_path (forgot exact name) in checkpoint be? Can these checkpoints be properly loaded and resumed? I'm just trying to look for edge cases here. 

@ananthsub @carmocca on the setting top_k=-1, yes it would save all checkpoints, but it is an important detail that the name "topk" would imply that we have a sorting based on the monitor. So even with top_k=-1 the monitor would be necessary. So @carmocca 's proposal with None does make sense to me at least for now :)

**Comment by carmocca** [comment]
>  what will the value of best_model_path (forgot exact name) in checkpoint be?

https://github.com/PyTorchLightning/pytorch-lightning/blob/f05244e036c0cf5c81623d4689d5a12815cab7cc/tests/callbacks/test_model_checkpoint.py#L219-L223
